File: examples/python/chat_core.py
"""Crypto core for the chatbot example bot.

A thin, network-free wrapper around the ``chat_xdk`` binding. Everything
that touches the SDK lives here so it can be unit-tested directly (see
``tests/test_chat_core.py``). The bot's I/O layer (``x_api.py``) and main loop
(``bot.py``) call into this module.

The four core feature touchpoints of the chat-xdk flow are all here:

* key management        -> ``load_keys`` / ``generate_and_register``
* conversation keys     -> ``prepare_conversation_key_change`` / ``decrypt_conversation_key``
* message encryption    -> ``encrypt_reply``
* event decryption      -> ``decrypt_batch`` (decrypt_events) and
                           ``decrypt_one`` (decrypt_event)
"""

from __future__ import annotations

import base64
from typing import Any

from chat_xdk import Chat


def _b64decode(data: str) -> bytes:
    return base64.b64decode(data)


def _as_dict(obj: Any) -> dict[str, Any]:
    """Decrypted events come back as native objects; normalise to a dict."""
    if isinstance(obj, dict):
        return obj
    if hasattr(obj, "model_dump"):
        return obj.model_dump()
    try:
        return dict(obj)
    except Exception:
        return {}


class ChatCore:
    """Wraps a single unlocked ``chat_xdk.Chat`` instance for one bot identity."""

    def __init__(self, chat: Chat | None = None) -> None:
        self.chat = chat or Chat()
        # The version the X API reports for our registered public key. Set after
        # loading/generating keys; used as the signing key version when sending.
        self.signing_key_version: str = "1"

    # -- Key management -----------------------------------------------------

    def load_keys(self, private_keys_b64: str, signing_key_version: str = "1") -> None:
        """Import an existing private-key blob (identity[+signing]) and unlock.

        ``private_keys_b64`` is the base64 of 32 bytes (identity only) or 64
        bytes (identity + signing) — the same blob ``export_keys`` produces.

        This is the quickstart key-storage path. For production key storage see
        ``unlock`` / ``setup`` (Juicebox).
        """
        self.chat.import_keys(_b64decode(private_keys_b64), version=signing_key_version)
        self.signing_key_version = signing_key_version

    def unlock(self, juicebox_config_json: str, pin: str, signing_key_version: str = "1") -> None:
        """Production key storage: recover the private keys from Juicebox.

        Use this instead of ``load_keys`` when keys are stored in Juicebox rather
        than a local blob. ``juicebox_config_json`` comes from the X API
        public-keys response; ``pin`` is the user's recovery PIN. Nothing secret
        is written to disk.
        """
        self.chat = Chat(juicebox_config_json)
        self.chat.unlock(pin)
        self.signing_key_version = signing_key_version

    def set_identity(self, user_id: str) -> None:
        """Set the session identity once the bot's own user id is known.

        Every signed action after this resolves its sender and signing-key
        version from the session, so the encrypt calls stay short.
        """
        self.chat.set_identity(user_id, self.signing_key_version)

    def set_signing_keys(self, signing_keys: list[dict[str, str]]) -> None:
        """Store participants' signing keys for decrypt calls that omit them.

        Each call replaces the previous set, so pass the full known roster.
        """
        self.chat.set_signing_keys(signing_keys)

    def set_cache_keys(self, enabled: bool) -> None:
        """Opt in to the SDK's verified conversation-key cache.

        While enabled, ``decrypt_batch`` feeds the cache and the encrypt
        methods can omit the conversation key pair entirely.
        """
        self.chat.set_cache_keys(enabled)

    def setup(self, juicebox_config_json: str, pin: str) -> dict[str, Any]:
        """First-time Juicebox provisioning: generate keys and store them under a PIN.

        Returns the registration payload to register the public key with the X
        API (see ``generate_and_register``). After this, future sessions recover
        the keys with ``unlock`` — no local blob needed.
        """
        self.chat = Chat(juicebox_config_json)
        payload = self.chat.generate_keypairs()
        self.chat.setup(pin)
        return {
            "public_key": payload.public_key.public_key,
            "signing_public_key": payload.public_key.signing_public_key,
            "identity_public_key_signature": payload.public_key.identity_public_key_signature,
            "registration_method": payload.public_key.registration_method,
        }

    def generate_and_register(self) -> dict[str, Any]:
        """Generate fresh keypairs for a brand-new bot identity.

        Returns the registration payload to POST to ``POST /2/users/public_keys``
        and the exported private-key blob to persist locally (e.g. in ``.env``).
        """
        payload = self.chat.generate_keypairs()
        exported = self.chat.export_keys()
        return {
            "registration": {
                "public_key": payload.public_key.public_key,
                "signing_public_key": payload.public_key.signing_public_key,
                "identity_public_key_signature": payload.public_key.identity_public_key_signature,
                "registration_method": payload.public_key.registration_method,
            },
            "version": payload.version,
            "private_keys_b64": base64.b64encode(bytes(exported)).decode("ascii")
            if exported
            else "",
        }

    def public_keys(self) -> dict[str, str]:
        keys = self.chat.get_public_keys()
        return {"identity": keys.identity, "signing": keys.signing}

    # -- Conversation keys --------------------------------------------------

    def prepare_conversation_key_change(
        self,
        public_keys: list[dict[str, str]],
        conversation_id: str | None = None,
    ) -> dict[str, Any]:
        """Generate, encrypt, and sign a conversation-key change.

        ``public_keys`` is the flat list returned by the X API's public-keys
        endpoint: ``[{"user_id", "public_key", "key_version"}, ...]``. Omit
        ``conversation_id`` for a one-to-one to derive it; pass it for a group.
        The change is signed with the ``set_identity`` session identity.
        """
        return self.chat.prepare_conversation_key_change(
            public_keys, conversation_id=conversation_id
        )

    def decrypt_conversation_key(self, encrypted_key_b64: str) -> bytes:
        """ECIES-decrypt one conversation key with our identity private key."""
        return self.chat.decrypt_conversation_key(encrypted_key_b64)

    # -- Decryption: the two paths -----------------------------------------

    def decrypt_batch(
        self,
        events_b64: list[str],
        signing_keys: list[dict[str, str]] | None = None,
    ) -> dict[str, Any]:
        """Batch path — used on initial conversation load.

        ``decrypt_events`` extracts conversation keys from any KeyChange events
        in the batch, then decrypts every message. Returns the decrypted
        messages plus the conversation-key cache to reuse for single events.
        Omit ``signing_keys`` to verify with the ``set_signing_keys`` store.
        """
        result = self.chat.decrypt_events(events_b64, signing_keys)
        messages = [
            {"event": _as_dict(m["event"]), "original_b64": m.get("original_b64")}
            for m in result.get("messages", [])
        ]
        return {
            "messages": messages,
            "conversation_keys": result.get("conversation_keys", {}),
            "errors": result.get("errors", {}),
        }

    def decrypt_one(
        self,
        event_b64: str,
        conversation_keys: dict[str, bytes],
        signing_keys: list[dict[str, str]] | None = None,
    ) -> dict[str, Any]:
        """Single-event path. ``conversation_keys`` is the cache from ``decrypt_batch``.

        Omit ``signing_keys`` to verify with the ``set_signing_keys`` store.
        """
        return _as_dict(self.chat.decrypt_event(event_b64, conversation_keys, signing_keys))

    # -- Message encryption -------------------------------------------------

    def encrypt_reply(
        self,
        *,
        conversation_id: str,
        text: str,
        conversation_key: bytes | None = None,
        conversation_key_version: str | None = None,
        reply_to_event: str | None = None,
        reply_to_sequence_id: str | None = None,
        entities: list[tuple[int, int, str]] | None = None,
        attachments: list[dict[str, Any]] | None = None,
        ttl_msec: int | None = None,
    ) -> dict[str, str]:
        """Encrypt + sign a message, returning fields ready for the X API send.

        Without a reply target this sends a fresh message via
        ``encrypt_message``; with ``reply_to_event`` (the raw encoded event
        being answered — preferred, the SDK derives and embeds the validated
        preview) or ``reply_to_sequence_id``, the SDK's ``encrypt_reply``
        builds a *threaded* reply. Omit the conversation key pair to resolve
        it from the SDK's verified-key cache (``set_cache_keys``).
        ``entities`` are ``(start, end, type)`` byte ranges; ``attachments``
        are attachment descriptors (e.g. a media reference); ``ttl_msec``
        makes the message disappear after the given lifetime. Signed with the
        ``set_identity`` session identity.
        """
        if reply_to_event is None and reply_to_sequence_id is None:
            payload = self.chat.encrypt_message(
                conversation_id,
                text,
                conversation_key=conversation_key,
                conversation_key_version=conversation_key_version,
                entities=entities,
                attachments=attachments,
                ttl_msec=ttl_msec,
            )
        else:
            payload = self.chat.encrypt_reply(
                conversation_id,
                text,
                reply_to_event,
                reply_to_sequence_id=reply_to_sequence_id,
                conversation_key=conversation_key,
                conversation_key_version=conversation_key_version,
                entities=entities,
                attachments=attachments,
                ttl_msec=ttl_msec,
            )
        return {
            # The SDK generates the message id and returns it in the payload.
            "message_id": payload.message_id,
            "encoded_message_create_event": payload.encrypted_content,
            "encoded_message_event_signature": payload.encoded_event_signature,
        }

    def encrypt_reaction(
        self,
        *,
        add: bool,
        conversation_id: str,
        target_message_sequence_id: str,
        emoji: str,
        conversation_key: bytes,
        conversation_key_version: str,
    ) -> dict[str, str]:
        """Encrypt + sign a reaction add/remove targeting a message sequence id.

        Signed with the ``set_identity`` session identity.
        """
        method = self.chat.encrypt_add_reaction if add else self.chat.encrypt_remove_reaction
        payload = method(
            None,
            emoji,
            conversation_id=conversation_id,
            target_message_sequence_id=target_message_sequence_id,
            conversation_key=conversation_key,
            conversation_key_version=conversation_key_version,
        )
        return {
            # The SDK generates the message id and returns it in the payload.
            "message_id": payload.message_id,
            "encoded_message_create_event": payload.encrypted_content,
            "encoded_message_event_signature": payload.encoded_event_signature,
        }

    # -- Group management -----------------------------------------------------

    def prepare_group_create(
        self,
        public_keys: list[dict[str, str]],
        conversation_id: str,
        member_ids: list[str],
        admin_ids: list[str],
    ) -> dict[str, Any]:
        """Prepare a group creation: fresh key + the two required signatures.

        Signed with the ``set_identity`` session identity.
        """
        return self.chat.prepare_group_create(
            public_keys,
            conversation_id,
            member_ids,
            admin_ids,
        )

    def prepare_group_members_change(
        self,
        public_keys: list[dict[str, str]],
        conversation_id: str,
        new_member_ids: list[str],
        current_member_ids: list[str],
        current_admin_ids: list[str],
    ) -> dict[str, Any]:
        """Prepare a member add: rotated key + the two required signatures.

        Signed with the ``set_identity`` session identity.
        """
        return self.chat.prepare_group_members_change(
            public_keys,
            conversation_id,
            new_member_ids,
            current_member_ids,
            current_admin_ids,
            [],
        )

    # -- Media streaming -----------------------------------------------------

    MEDIA_CHUNK = 1024 * 1024

    def encrypt_media(self, plaintext: bytes, conversation_key: bytes) -> bytes:
        """Encrypt a media blob with the incremental stream API.

        Feeding fixed-size chunks through ``push`` keeps memory bounded no
        matter how large the file is; ``finish`` emits the final frame that
        seals the stream (decryption fails without it).
        """
        enc = self.chat.stream_encryptor(conversation_key)
        out = bytearray()
        for offset in range(0, len(plaintext), self.MEDIA_CHUNK):
            out += enc.push(plaintext[offset : offset + self.MEDIA_CHUNK])
        out += enc.finish()
        return bytes(out)

    def decrypt_media(self, ciphertext: bytes, conversation_key: bytes) -> bytes:
        """Decrypt a media blob with the incremental stream API.

        ``finish`` raises if the stream was truncated, so plaintext from
        ``push`` must not be treated as complete until it succeeds.
        """
        dec = self.chat.stream_decryptor(conversation_key)
        out = bytearray()
        for offset in range(0, len(ciphertext), self.MEDIA_CHUNK):
            out += dec.push(ciphertext[offset : offset + self.MEDIA_CHUNK])
        out += dec.finish()
        return bytes(out)

    # -- Generic helpers (handy for metadata + tests) -----------------------

    def encrypt(self, plaintext: str, conversation_key: bytes) -> str:
        return self.chat.encrypt(plaintext, conversation_key)

    def decrypt(self, ciphertext_b64: str, conversation_key: bytes) -> str:
        return self.chat.decrypt(ciphertext_b64, conversation_key)


def message_text(event: dict[str, Any]) -> str | None:
    """Pull the plain text out of a decrypted Message event, or None."""
    if event.get("type") != "Message":
        return None
    content = event.get("content") or {}
    return content.get("text")


def prep_to_request(prep: dict[str, Any], signing_public_key: str) -> dict[str, Any]:
    """Map a prepared conversation change into the X API request shape.

    Works for 1:1 key changes (one signature) and group create / member add
    (two signatures). ``signing_public_key`` is the sender's own signing key,
    which the API expects alongside each signature.
    """
    body: dict[str, Any] = {
        "conversation_key_version": prep["conversation_key_version"],
        "conversation_participant_keys": [
            {
                "user_id": str(pk["user_id"]),
                "encrypted_conversation_key": pk["encrypted_key"],
                "public_key_version": str(pk["public_key_version"]),
            }
            for pk in prep["participant_keys"]
        ],
        "action_signatures": [
            {
                "message_id": sig["message_id"],
                "encoded_message_event_detail": sig["encoded_message_event_detail"],
                "message_event_signature": {
                    "signature": sig["signature"],
                    "signature_version": sig["signature_version"],
                    "public_key_version": sig["public_key_version"],
                    "signing_public_key": signing_public_key,
                },
                **(
                    {"signature_payload": sig["signature_payload"]}
                    if sig.get("signature_payload")
                    else {}
                ),
            }
            for sig in prep["action_signatures"]
        ],
    }
    return body

File: examples/js/src/chat-core.mjs
/**
 * Crypto core for the JavaScript/WASM chat-xdk example.
 *
 * A thin wrapper around the chat-xdk WASM binding. Everything that touches the
 * SDK lives here so it can be unit-tested directly (see
 * test/chat-core.test.mjs). The four core feature touchpoints are all here:
 *
 *   - key management     -> loadKeys / generateAndRegister / setIdentity
 *   - conversation keys  -> prepareConversationKeyChange / setCacheKeys
 *   - message encryption -> encryptMessage / encryptReply / encryptReaction
 *   - event decryption   -> decryptBatch (decryptEvents) and decryptOne (decryptEvent)
 *
 * The session model: identity (setIdentity), signing keys (setSigningKeys),
 * and the conversation-key cache (setCacheKeys) are set once, so the
 * per-message calls need only the conversation id and text.
 */
import { readFile } from "node:fs/promises";

import init, {
  Chat,
  base64ToBytes,
  bytesToBase64,
} from "../../../crates/wasm/pkg/chat_xdk_wasm.js";

// The WASM public API does not expose raw key import/export, so the local-blob
// example drives the lower-level WASM Chat directly.
let wasmReady;
/** Concatenate Uint8Array chunks into one buffer. */
function concatBytes(parts) {
  const out = new Uint8Array(parts.reduce((n, p) => n + p.length, 0));
  let offset = 0;
  for (const p of parts) {
    out.set(p, offset);
    offset += p.length;
  }
  return out;
}

function ensureWasm() {
  if (!wasmReady) {
    const wasmUrl = new URL(
      "../../../crates/wasm/pkg/chat_xdk_wasm_bg.wasm",
      import.meta.url,
    );
    wasmReady = readFile(wasmUrl).then((bytes) => init({ module_or_path: bytes }));
  }
  return wasmReady;
}

export class ChatCore {
  #chat;
  signingKeyVersion = "1";

  constructor(chat) {
    this.#chat = chat;
  }

  /** Async factory — the WASM module must be initialized before use. */
  static async create() {
    await ensureWasm();
    return new ChatCore(new Chat());
  }

  // -- Key management -------------------------------------------------------

  /** Import an existing base64 private-key blob (identity[+signing]). */
  loadKeys(privateKeysB64, signingKeyVersion = "1") {
    const bytes = base64ToBytes(privateKeysB64);
    if (!bytes) throw new Error("invalid base64 private keys");
    this.#chat.importKeys(bytes, signingKeyVersion);
    this.signingKeyVersion = signingKeyVersion;
  }

  /** Generate a fresh identity; returns the registration payload + private blob. */
  generateAndRegister() {
    const payload = this.#chat.generateKeypairs();
    const exported = this.#chat.exportKeys();
    return { registration: payload, privateKeysB64: bytesToBase64(exported) };
  }

  publicKeys() {
    return this.#chat.getPublicKeys();
  }

  // -- Session --------------------------------------------------------------

  /**
   * Set the session identity once; every encrypt/prepare call afterwards
   * resolves its sender id and signing-key version from it.
   */
  setIdentity(userId) {
    this.#chat.setIdentity(userId, this.signingKeyVersion);
  }

  /**
   * Store the participants' signing keys; decrypt calls that omit their
   * signingKeys argument verify against this store.
   */
  setSigningKeys(signingKeys) {
    this.#chat.setSigningKeys(signingKeys);
  }

  /**
   * Opt in to the conversation-key cache: decryptBatch caches each
   * conversation's verified key, and encrypt calls that omit the key pair
   * resolve it from the cache.
   */
  setCacheKeys(enabled) {
    this.#chat.setCacheKeys(enabled);
  }

  // -- Conversation keys ----------------------------------------------------

  /**
   * Generate, encrypt, and sign a conversation-key change.
   *
   * `publicKeys` is `[{ userId, publicKey, keyVersion }, ...]`. Omit
   * `conversationId` for a one-to-one to derive it; pass it for a group.
   */
  prepareConversationKeyChange({ publicKeys, conversationId }) {
    return this.#chat.prepareConversationKeyChange({ publicKeys, conversationId });
  }

  /** ECIES-decrypt one conversation key -> raw 32-byte Uint8Array. */
  decryptConversationKey(encryptedKeyB64) {
    return this.#chat.decryptConversationKey(encryptedKeyB64);
  }

  // -- Decryption: the two paths -------------------------------------------

  /** Batch path — used on initial conversation load (also feeds the key cache). */
  decryptBatch(eventsB64, signingKeys) {
    return this.#chat.decryptEvents(eventsB64, signingKeys);
  }

  /**
   * Single-event path — used for each new event after the initial load.
   * With the session set up, both key arguments can be omitted:
   * conversation keys resolve from the cache and signing keys from the store.
   */
  decryptOne(eventB64, conversationKeys, signingKeys) {
    return this.#chat.decryptEvent(eventB64, conversationKeys, signingKeys);
  }

  // -- Message encryption ---------------------------------------------------

  /**
   * Encrypt + sign a message; returns fields ready for the X API send.
   * With `replyToEvent` (the base64 raw event being answered) the SDK builds
   * a *threaded* reply and derives the preview from the signed original;
   * without it this sends a fresh message. The conversation key pair is
   * optional once the cache is enabled. `entities` are `[start, end, type]`
   * byte ranges; `ttlMsec` makes the message disappear after the given
   * lifetime.
   */
  encryptReply({
    conversationId,
    text,
    conversationKey,
    conversationKeyVersion,
    replyToEvent,
    entities,
    attachments,
    ttlMsec,
  }) {
    const params = {
      conversationId,
      text,
      conversationKey,
      conversationKeyVersion,
      entities,
      attachments,
      ttlMsec,
    };
    const payload = replyToEvent
      ? this.#chat.encryptReply({ ...params, replyToEvent })
      : this.#chat.encryptMessage(params);
    return {
      // The SDK generates the message id and returns it in the payload.
      message_id: payload.messageId,
      encoded_message_create_event: payload.encryptedContent,
      encoded_message_event_signature: payload.encodedEventSignature,
    };
  }

  /**
   * Encrypt + sign a reaction add/remove. `targetEvent` is the base64 raw
   * event being reacted to; the conversation id and target sequence id are
   * derived from it.
   */
  encryptReaction({ add, targetEvent, emoji, conversationKey, conversationKeyVersion }) {
    const params = { emoji, targetEvent, conversationKey, conversationKeyVersion };
    const payload = add
      ? this.#chat.encryptAddReaction(params)
      : this.#chat.encryptRemoveReaction(params);
    return {
      // The SDK generates the message id and returns it in the payload.
      message_id: payload.messageId,
      encoded_message_create_event: payload.encryptedContent,
      encoded_message_event_signature: payload.encodedEventSignature,
    };
  }

  /**
   * Encrypt + sign a message edit. `targetEvent` is the base64 raw event of
   * the message being edited; the conversation id and target sequence id are
   * derived from it.
   */
  encryptEdit({ targetEvent, updatedText, entities, conversationKey, conversationKeyVersion }) {
    const payload = this.#chat.encryptEdit({
      targetEvent,
      updatedText,
      entities,
      conversationKey,
      conversationKeyVersion,
    });
    return {
      // The SDK generates the message id and returns it in the payload.
      message_id: payload.messageId,
      encoded_message_create_event: payload.encryptedContent,
      encoded_message_event_signature: payload.encodedEventSignature,
    };
  }

  /**
   * Sign a message delete. Returns the action signature to submit alongside
   * the delete request.
   */
  prepareMessageDelete({ conversationId, sequenceIds, deleteForAll }) {
    return this.#chat.prepareMessageDelete({ conversationId, sequenceIds, deleteForAll });
  }

  // -- Group management ------------------------------------------------------

  /** Prepare a group creation: fresh key + the two required signatures. */
  prepareGroupCreate({ publicKeys, conversationId, memberIds, adminIds }) {
    return this.#chat.prepareGroupCreate({
      publicKeys,
      conversationId,
      memberIds,
      adminIds,
    });
  }

  /** Prepare a member add: rotated key + the two required signatures. */
  prepareGroupMembersChange({
    publicKeys,
    conversationId,
    newMemberIds,
    currentMemberIds,
    currentAdminIds,
  }) {
    return this.#chat.prepareGroupMembersChange({
      publicKeys,
      conversationId,
      newMemberIds,
      currentMemberIds,
      currentAdminIds,
      currentPendingMemberIds: [],
    });
  }

  // -- Media streaming --------------------------------------------------------

  static MEDIA_CHUNK = 1024 * 1024;

  /**
   * Encrypt a media blob with the incremental stream API.
   *
   * Feeding fixed-size chunks through `push` keeps memory bounded no matter
   * how large the file is (the WASM heap cannot hold large files whole);
   * `finish` emits the final frame that seals the stream — decryption fails
   * without it.
   */
  encryptMedia(plaintext, conversationKey) {
    const enc = this.#chat.streamEncryptor(conversationKey);
    const parts = [];
    // A throw from push would otherwise leak the encryptor's live secretstream
    // key state in WASM memory; free() reclaims it. finish() consumes the
    // encryptor even when it fails, so it needs no explicit cleanup (and
    // free() after it would hit an already-null pointer).
    try {
      for (let offset = 0; offset < plaintext.length; offset += ChatCore.MEDIA_CHUNK) {
        parts.push(enc.push(plaintext.subarray(offset, offset + ChatCore.MEDIA_CHUNK)));
      }
    } catch (e) {
      enc.free();
      throw e;
    }
    parts.push(enc.finish());
    return concatBytes(parts);
  }

  /**
   * Decrypt a media blob with the incremental stream API.
   *
   * `finish` throws if the stream was truncated, so plaintext from `push`
   * must not be treated as complete until it succeeds.
   */
  decryptMedia(ciphertext, conversationKey) {
    const dec = this.#chat.streamDecryptor(conversationKey);
    const parts = [];
    // As in encryptMedia: free() the decryptor if push throws mid-stream;
    // finish() consumes it even when it fails.
    try {
      for (let offset = 0; offset < ciphertext.length; offset += ChatCore.MEDIA_CHUNK) {
        parts.push(dec.push(ciphertext.subarray(offset, offset + ChatCore.MEDIA_CHUNK)));
      }
    } catch (e) {
      dec.free();
      throw e;
    }
    parts.push(dec.finish());
    return concatBytes(parts);
  }

  // -- Generic helpers (handy for metadata + tests) -------------------------

  encrypt(plaintext, conversationKey) {
    return this.#chat.encrypt(plaintext, conversationKey);
  }

  decrypt(ciphertextB64, conversationKey) {
    return this.#chat.decrypt(ciphertextB64, conversationKey);
  }
}

/** Pull the plain text out of a decrypted Message event, or null. */
export function messageText(event) {
  if (!event || event.type !== "message") return null;
  return event.content?.text ?? null;
}

/**
 * Map a prepared conversation change into the X API request shape.
 *
 * Works for 1:1 key changes (one signature) and group create / member add
 * (two signatures). `signingPublicKey` is the sender's own signing key,
 * which the API expects alongside each signature.
 */
export function prepToRequest(prep, signingPublicKey) {
  return {
    conversationKeyVersion: prep.conversationKeyVersion,
    conversationParticipantKeys: prep.participantKeys.map((pk) => ({
      userId: String(pk.userId),
      encryptedConversationKey: pk.encryptedKey,
      publicKeyVersion: String(pk.publicKeyVersion),
    })),
    actionSignatures: prep.actionSignatures.map((sig) => ({
      messageId: sig.messageId,
      encodedMessageEventDetail: sig.encodedMessageEventDetail,
      ...(sig.signaturePayload ? { signaturePayload: sig.signaturePayload } : {}),
      messageEventSignature: {
        signature: sig.signature,
        signatureVersion: sig.signatureVersion,
        publicKeyVersion: sig.publicKeyVersion,
        signingPublicKey,
      },
    })),
  };
}
